File: accbot-android/app/src/main/java/com/accbot/dca/domain/usecase/CalculateMonthlyCostUseCase.kt
package com.accbot.dca.domain.usecase

import com.accbot.dca.data.remote.MarketDataService
import com.accbot.dca.domain.model.DcaFrequency
import com.accbot.dca.domain.model.DcaStrategy
import com.accbot.dca.domain.util.CronUtils
import com.accbot.dca.presentation.model.MonthlyCostEstimate
import java.math.BigDecimal
import java.math.RoundingMode
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CalculateMonthlyCostUseCase @Inject constructor(
    private val marketDataService: MarketDataService
) {
    fun getEffectiveIntervalMinutes(frequency: DcaFrequency, cronExpression: String?): Long {
        return if (frequency == DcaFrequency.CUSTOM) {
            CronUtils.getIntervalMinutesEstimate(cronExpression ?: "") ?: 1440L
        } else {
            frequency.intervalMinutes
        }
    }

    suspend fun computeEstimate(
        amount: BigDecimal,
        frequency: DcaFrequency,
        cronExpression: String?,
        strategy: DcaStrategy,
        crypto: String,
        fiat: String
    ): MonthlyCostEstimate? {
        if (amount <= BigDecimal.ZERO) return null

        val intervalMinutes = getEffectiveIntervalMinutes(frequency, cronExpression)
        if (intervalMinutes <= 0) return null

        val runsPerMonth = BigDecimal(30 * 24 * 60).divide(
            BigDecimal(intervalMinutes), 2, RoundingMode.HALF_UP
        )

        return when (strategy) {
            is DcaStrategy.Classic -> {
                val monthly = amount.multiply(runsPerMonth)
                MonthlyCostEstimate(
                    minMonthly = monthly,
                    maxMonthly = monthly,
                    currentMonthly = monthly,
                    currentInfo = null
                )
            }
            is DcaStrategy.AthBased -> {
                computeStrategyEstimate(
                    amount = amount,
                    runsPerMonth = runsPerMonth,
                    minMult = strategy.tiers.minOf { it.multiplier },
                    maxMult = strategy.tiers.maxOf { it.multiplier },
                    fetchCurrentMultiplier = {
                        val cryptoData = marketDataService.getCryptoData(crypto, fiat)
                        if (cryptoData != null) {
                            val mult = strategy.tiers.sortedBy { it.maxDistancePercent }
                                .firstOrNull { cryptoData.athDistance <= it.maxDistancePercent }
                                ?.multiplier ?: 1.0f
                            mult to "$crypto is ${cryptoData.athDistancePercent}% below ATH"
                        } else null
                    }
                )
            }
            is DcaStrategy.FearAndGreed -> {
                computeStrategyEstimate(
                    amount = amount,
                    runsPerMonth = runsPerMonth,
                    minMult = strategy.tiers.minOf { it.multiplier },
                    maxMult = strategy.tiers.maxOf { it.multiplier },
                    fetchCurrentMultiplier = {
                        val fngData = marketDataService.getFearGreedIndex()
                        if (fngData != null) {
                            val mult = strategy.tiers.sortedBy { it.maxIndex }
                                .firstOrNull { fngData.value <= it.maxIndex }
                                ?.multiplier ?: 1.0f
                            mult to "Fear & Greed: ${fngData.value} (${fngData.classification})"
                        } else null
                    }
                )
            }
        }
    }

    private suspend fun computeStrategyEstimate(
        amount: BigDecimal,
        runsPerMonth: BigDecimal,
        minMult: Float,
        maxMult: Float,
        fetchCurrentMultiplier: suspend () -> Pair<Float, String>?
    ): MonthlyCostEstimate {
        val minMonthly = amount.multiply(BigDecimal(minMult.toString())).multiply(runsPerMonth)
        val maxMonthly = amount.multiply(BigDecimal(maxMult.toString())).multiply(runsPerMonth)

        val result = fetchCurrentMultiplier()
        val currentMonthly = result?.let { (mult, _) ->
            amount.multiply(BigDecimal(mult.toString())).multiply(runsPerMonth)
        }

        return MonthlyCostEstimate(
            minMonthly = minMonthly,
            maxMonthly = maxMonthly,
            currentMonthly = currentMonthly,
            currentInfo = result?.second
        )
    }
}

File: accbot-ios/AccBot/Domain/UseCase/CalculateMonthlyCostUseCase.swift
import Foundation

/// Monthly cost estimate with range for dynamic strategies.
struct MonthlyCostEstimate {
    let minMonthly: Decimal
    let maxMonthly: Decimal
    let currentMonthly: Decimal?
    let currentInfo: String?
}

/// Estimates monthly DCA cost based on amount, frequency, and strategy.
final class CalculateMonthlyCostUseCase {
    private let marketDataService: MarketDataService

    init(marketDataService: MarketDataService) {
        self.marketDataService = marketDataService
    }

    func getEffectiveIntervalMinutes(frequency: DcaFrequency, cronExpression: String?) -> Int {
        if frequency == .custom {
            return CronUtils.getIntervalMinutesEstimate(cron: cronExpression ?? "") ?? 1440
        }
        return frequency.intervalMinutes
    }

    func computeEstimate(
        amount: Decimal,
        frequency: DcaFrequency,
        cronExpression: String?,
        strategy: DcaStrategy,
        crypto: String,
        fiat: String
    ) async -> MonthlyCostEstimate? {
        guard amount > 0 else { return nil }

        let intervalMinutes = getEffectiveIntervalMinutes(frequency: frequency, cronExpression: cronExpression)
        guard intervalMinutes > 0 else { return nil }

        let minutesPerMonth = Decimal(30 * 24 * 60)
        let runsPerMonth = divideDecimal(minutesPerMonth, by: Decimal(intervalMinutes), scale: 2)

        switch strategy {
        case .classic:
            let monthly = amount * runsPerMonth
            return MonthlyCostEstimate(minMonthly: monthly, maxMonthly: monthly, currentMonthly: monthly, currentInfo: nil)

        case .athBased(let tiers):
            return await computeStrategyEstimate(
                amount: amount,
                runsPerMonth: runsPerMonth,
                minMult: tiers.map(\.multiplier).min() ?? 0.5,
                maxMult: tiers.map(\.multiplier).max() ?? 3.0
            ) {
                guard let currentPrice = await self.marketDataService.getCurrentPrice(crypto: crypto, fiat: fiat),
                      let ath = await self.marketDataService.getAllTimeHigh(crypto: crypto, fiat: fiat),
                      ath > 0
                else { return nil }

                let athDistance = Float(1.0 - NSDecimalNumber(decimal: currentPrice).doubleValue / NSDecimalNumber(decimal: ath).doubleValue)
                let distPct = Int(athDistance * 100)
                let mult = tiers.sorted(by: { $0.maxDistancePercent < $1.maxDistancePercent })
                    .first(where: { athDistance <= $0.maxDistancePercent })?.multiplier ?? 1.0
                return (mult, "\(crypto) is \(distPct)% below ATH")
            }

        case .fearAndGreed(let tiers):
            return await computeStrategyEstimate(
                amount: amount,
                runsPerMonth: runsPerMonth,
                minMult: tiers.map(\.multiplier).min() ?? 0.25,
                maxMult: tiers.map(\.multiplier).max() ?? 2.5
            ) {
                guard let index = await self.marketDataService.getFearGreedIndex() else { return nil }
                let mult = tiers.sorted(by: { $0.maxIndex < $1.maxIndex })
                    .first(where: { index <= $0.maxIndex })?.multiplier ?? 1.0
                let classification: String = switch index {
                case 0...24: "Extreme Fear"
                case 25...44: "Fear"
                case 45...54: "Neutral"
                case 55...74: "Greed"
                default: "Extreme Greed"
                }
                return (mult, "Fear & Greed: \(index) (\(classification))")
            }
        }
    }

    private func computeStrategyEstimate(
        amount: Decimal,
        runsPerMonth: Decimal,
        minMult: Float,
        maxMult: Float,
        fetchCurrentMultiplier: () async -> (Float, String)?
    ) async -> MonthlyCostEstimate {
        let minMonthly = amount * Decimal(Double(minMult)) * runsPerMonth
        let maxMonthly = amount * Decimal(Double(maxMult)) * runsPerMonth

        let result = await fetchCurrentMultiplier()
        let currentMonthly = result.map { amount * Decimal(Double($0.0)) * runsPerMonth }

        return MonthlyCostEstimate(
            minMonthly: minMonthly,
            maxMonthly: maxMonthly,
            currentMonthly: currentMonthly,
            currentInfo: result?.1
        )
    }

    private func divideDecimal(_ a: Decimal, by b: Decimal, scale: Int) -> Decimal {
        let handler = NSDecimalNumberHandler(
            roundingMode: .plain, scale: Int16(scale),
            raiseOnExactness: false, raiseOnOverflow: false,
            raiseOnUnderflow: false, raiseOnDivideByZero: false
        )
        return NSDecimalNumber(decimal: a)
            .dividing(by: NSDecimalNumber(decimal: b), withBehavior: handler)
            .decimalValue
    }
}
